romanisim/tests/test_linear.py

```diff
@@ -0,0 +1,185 @@
+"""Unit tests for linear module.
+
+Routines tested:
+- validate_times
+- tij_to_pij
+- apportion_counts_to_resultants
+- add_read_noise_to_resultants
+- make_asdf
+- ma_table_to_tij
+- make_l1
+"""
+import os
+import pytest
+import numpy as np
+import galsim
+import copy
+from astropy import units as u
+import crds
+
+import roman_datamodels
+import roman_datamodels.maker_utils as maker_utils
+from romanisim import nonlinearity, parameters, wcs, catalog, image, util
+
+
+
+tijlist = [
+    [[1], [2, 3], [4, 5], [6, 7]],
+    [[100], [101, 102, 103], [110]],
+    [[1], [2], [3, 4, 5, 100]],
+]
+
+ma_table_list = [
+    [[1, 10], [11, 1], [12, 10], [30, 1], [40, 5], [50, 100]],
+    [[1, 1]],
+    [[1, 1], [10, 1]],
+]
+
+INVERSE = [True, False]
+
+
+def test_linear_apply():
+    counts = np.random.poisson(100, size=(100, 100))
+    coeffs = np.asfarray([1.0, 0.7, 3.0e-6, 5.0e-12])
+    lin_coeffs = np.tile(coeffs[:, np.newaxis, np.newaxis], (1, 100, 100))
+    lin_coeffs[:, 0:50, :] *= 2.0
+    gain = 4.0 * u.electron / u.DN
+    counts[0, 0] = counts[0, 99] = counts[99, 0] = counts[99, 99] = 11.0
+
+    linearity = nonlinearity.NL(lin_coeffs, gain)
+
+    res2 = linearity.apply(counts, electrons=True)
+
+    assert res2[0,0] == res2[0,99]
+    assert res2[99, 0] == res2[99, 99]
+    assert res2[0, 0] == 2 * res2[99, 99]
+
+
+@pytest.mark.parametrize("inverse", INVERSE)
+def test_repair_coeffs(inverse):
+    counts = np.random.poisson(100, size=(100, 100))
+    coeffs = np.asfarray([1.0, 0.7, 3.0e-6, 5.0e-12])
+    lin_coeffs = np.tile(coeffs[:, np.newaxis, np.newaxis], (1, 100, 100))
+    lin_coeffs[:, 0:50, :] *= 2.0
+    lin_coeffs[:, 1, 1] *= 0
+    lin_coeffs[2, 22, 22] = np.nan
+    gain = 4.0 * u.electron / u.DN
+
+    linearity = nonlinearity.NL(lin_coeffs, gain, inverse)
+
+    res = linearity.apply(counts)
+
+    assert res[1, 1] == counts[1, 1]
+    assert res[22, 22] == counts[22, 22]
+    assert np.all(res[23:,23:] != counts[23:,23:])
```

Comment:
This is really "all of the array except the two pixels I messed with"?  Maybe np.sum(res != counts) == np.prod(counts.shape) - 2 plus a comment?

Reply:
Changed.